cdn/signUrl/SignUrlSample/Program.cs

```diff
@@ -0,0 +1,119 @@
+        public DateTime Expiration { get; set; }
+    }
+
+    public class Program
+    {
+        static int Main(string[] args)
+        {
+            string signedUrl = null;
+            Parser.Default.ParseArguments<SignedUrlArgs>(args).MapResult(
+                uargs => signedUrl = CreateSignedUrl(uargs.Url, uargs.KeyName, File.ReadAllText(uargs.KeyPath, Encoding.UTF8), uargs.Expiration), 
+                err => string.Empty);
+
+            if (string.IsNullOrEmpty(signedUrl))
+            {
+                return 1;
+            }
+
+            Console.WriteLine($"Signed URL: {signedUrl}");
+            return 0;
+        }
+
+        // [START example]
+        /// <summary>
+        /// Creates signed URL for Google Cloud SDN
+        /// More details about order of operations is here: <see cref="https://cloud-dot-devsite.googleplex.com/cdn/docs/signed-urls#creatingkeys"/>
+        /// </summary>
+        /// <param name="url">The Url to sign. This URL can't include Expires and KeyName query parameters in it</param>
+        /// <param name="keyName">The name of the key used to sign the URL</param>
+        /// <param name="encodedKey">The key used to sign the Url</param>
+        /// <param name="expirationTime">Expiration time of the signature</param>
+        /// <returns>Signed Url that is valid until {expirationTime}</returns>
+        public static string CreateSignedUrl(string url, string keyName, string encodedKey, DateTime expirationTime)
+        {
+            var builder = new UriBuilder(url);
+
+            long unixTimestampExpiration = ToUnixTime(expirationTime);
+
+            char queryParam = string.IsNullOrEmpty(builder.Query) ? '?' : '&';
+            builder.Query += $"{queryParam}Expires={unixTimestampExpiration}&KeyName={keyName}".ToString();
+
+            // Key is passed as base64url encoded
+            byte[] decodedKey = Base64UrlDecode(encodedKey);
+
+            // Computes HMAC SHA-1 hash of the URL using the key
+            byte[] hash = ComputeHash(decodedKey, builder.Uri.AbsoluteUri);
+            string encodedHash = Base64UrlEncode(hash);
+            
+            builder.Query += $"&Signature={encodedHash}";
+            return builder.Uri.AbsoluteUri;
+        }
+
+        private static long ToUnixTime(DateTime date)
+        {
+            var epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+            return Convert.ToInt64((date - epoch).TotalSeconds);
+        }
+
+        private static byte[] Base64UrlDecode(string arg)
+        {
+            string s = arg;
+            s = s.Replace('-', '+'); // 62nd char of encoding
+            s = s.Replace('_', '/'); // 63rd char of encoding
+
+            return Convert.FromBase64String(s); // Standard base64 decoder
+        }
+
+        private static string Base64UrlEncode(byte[] inputBytes)
+        {
+            var output = Convert.ToBase64String(inputBytes);
```

Comment:
not sure why this entire method cannot be

```
return Convert.ToBase64String(inputBytes).Replace('+', '-')
                           .Replace('/', '_');
```

Reply:
To me, it's 6 vs. half dozen.

Reply:
@SurferJeffAtGoogle , not sure what do you mean by your comment. Can you explain?